src/anndata/_io/specs/methods.py

```diff
@@ -1028,30 +1034,82 @@ def write_nullable_integer(
+    v: pd.arrays.StringArray,
+    *,
+    _writer: Writer,
+    dataset_kwargs: Mapping[str, Any] = MappingProxyType({}),
+):
+    from ..._settings import settings
+
+    if not settings.allow_write_nullable_strings:
+        msg = (
+            "`anndata.settings.allow_write_nullable_strings` is False, "
+            "because writing of `pd.arrays.StringArray` is new "
+            "and not supported in anndata < 0.11, still use by many people. "
+            "Opt-in to writing these arrays by toggling the setting to True."
+        )
+        raise RuntimeError(msg)
+    g = f.require_group(k)
+    if (mask := v.isna()).any():
+        _writer.write_elem(g, "mask", mask, dataset_kwargs=dataset_kwargs)
+    _writer.write_elem(
+        g, "values", v.fillna("")._ndarray, dataset_kwargs=dataset_kwargs
+    )
+
+
 @_REGISTRY.register_read(H5Group, IOSpec("nullable-integer", "0.1.0"))
 @_REGISTRY.register_read(ZarrGroup, IOSpec("nullable-integer", "0.1.0"))
 def read_nullable_integer(
     elem: GroupStorageType, *, _reader: Reader
 ) -> pd.api.extensions.ExtensionArray:
-    if "mask" in elem:
-        return pd.arrays.IntegerArray(
-            _reader.read_elem(elem["values"]), mask=_reader.read_elem(elem["mask"])
-        )
-    else:
-        return pd.array(_reader.read_elem(elem["values"]))
+    return _read_nullable(elem, _reader=_reader, array_type=pd.arrays.IntegerArray)
 
 
 @_REGISTRY.register_read(H5Group, IOSpec("nullable-boolean", "0.1.0"))
 @_REGISTRY.register_read(ZarrGroup, IOSpec("nullable-boolean", "0.1.0"))
 def read_nullable_boolean(
     elem: GroupStorageType, *, _reader: Reader
 ) -> pd.api.extensions.ExtensionArray:
-    if "mask" in elem:
-        return pd.arrays.BooleanArray(
-            _reader.read_elem(elem["values"]), mask=_reader.read_elem(elem["mask"])
-        )
-    else:
-        return pd.array(_reader.read_elem(elem["values"]))
+    return _read_nullable(elem, _reader=_reader, array_type=pd.arrays.BooleanArray)
+
+
+@_REGISTRY.register_read(H5Group, IOSpec("nullable-string-array", "0.1.0"))
+@_REGISTRY.register_read(ZarrGroup, IOSpec("nullable-string-array", "0.1.0"))
+def read_nullable_string(
+    elem: GroupStorageType, *, _reader: Reader
+) -> pd.api.extensions.ExtensionArray:
+    def string_array(
+        values: np.ndarray, mask: np.ndarray
+    ) -> pd.api.extensions.ExtensionArray:
+        arr = pd.array(values, dtype="string")
+        arr[mask] = pd.NA
+        return arr
+
+    return _read_nullable(elem, _reader=_reader, array_type=string_array)
+
+
+def _read_nullable(
+    elem: GroupStorageType,
+    *,
+    _reader: Reader,
+    # BaseMaskedArray
+    array_type: Callable[
+        [NDArray[np.number], NDArray[np.bool_]], pd.api.extensions.ExtensionArray
+    ],
+) -> pd.api.extensions.ExtensionArray:
+    values = _reader.read_elem(elem["values"])
+    if "mask" not in elem:
+        return pd.array(values)
+    return array_type(values, mask=_reader.read_elem(elem["mask"]))
```

Comment:
Would really like to understand what is going on when you have no `mask` i.e., how this could occur

Reply:
When the `pd.*.BaseMaskedArray`’s `_mask` is `None`:

https://github.com/scverse/anndata/blob/3afc34b948a6c16752dbf0c34a135951fb0664ea/src/anndata/_io/specs/methods.py#L1033-L1034